diff --git a/components/MarkdownContent.astro b/components/MarkdownContent.astro
index 0d40cd47ed269a706b05242f0c7994ed2a95a43d..12af4519f54ea1115168106862eaf6ab67d35eb8 100644
--- a/components/MarkdownContent.astro
+++ b/components/MarkdownContent.astro
@@ -4,3 +4,95 @@ import '../style/markdown.css';
 ---
 
 <div class="sl-markdown-content"><slot /></div>
+
+{
+	/* This is intentionally inlined to avoid FOUC. It lives here and not in
+	TableOfContents.astro because because it needs access to the complete document
+	outline. Setting a global function in TableOfContents.astro and invoking it
+	here results in FOUC on Safari. */
+}
+<script is:inline>
+	(() => {
+		// Table of Contents Selection
+		const toc = document.querySelector('starlight-toc nav');
+		const mobileToc = document.querySelector('mobile-starlight-toc nav');
+
+		if (toc && mobileToc) {
+			const tocLinks = [...toc.querySelectorAll('a')];
+			const mobileTocLinks = [...mobileToc.querySelectorAll('a')];
+			const headings = tocLinks.map((link) => {
+				const id = link.getAttribute('href')?.slice(1);
+				return document.getElementById(id || '');
+			});
+
+			// cache styles
+			const headingsMarginTop = headings.map((heading) => {
+				const style = window.getComputedStyle(heading);
+				return parseInt(style.marginTop, 10);
+			});
+
+			// `<summary>` only exists in mobile ToC, so will fall back to 0 in large viewport component.
+			const navBarHeight = document.querySelector('header')?.getBoundingClientRect().height || 0;
+			const mobileTocHeight = mobileToc.querySelector('summary')?.getBoundingClientRect().height || 0;
+			const activationThreshold = navBarHeight + mobileTocHeight;
+
+			let willScroll = false;
+
+			// find the id of the first heading that's in view
+			function findActiveHeadingId() {
+				for (let i = 0; i < headings.length; i++) {
+					const top = headings[i].getBoundingClientRect().top - headingsMarginTop[i];
+
+					if (top >= activationThreshold) {
+						return i === 0 ? headings[i].id : headings[i - 1].id;
+					}
+				}
+				return headings[headings.length - 1].id;
+			}
+
+			function activateTocLink(idOrHash) {
+				const id = idOrHash && idOrHash.startsWith('#') ? idOrHash.slice(1) : idOrHash;
+				let activeLink;
+				[...mobileTocLinks, ...tocLinks].forEach((link) => {
+					if (link.getAttribute('href')?.slice(1) === id) {
+						!link.hasAttribute('aria-current') && link.setAttribute('aria-current', 'true');
+						activeLink = link;
+					} else {
+						link.hasAttribute('aria-current') && link.removeAttribute('aria-current');
+					}
+				});
+
+				// update mobile ToC display
+				const display = mobileToc.querySelector('.display-current');
+				if (display) display.textContent = activeLink.textContent;
+			}
+
+			// special case, navigated to an anchor link
+			if (window.location.hash.length > 0) {
+				activateTocLink(window.location.hash);
+			} else {
+				activateTocLink(findActiveHeadingId());
+			}
+
+			// watch for hash changes
+			window.addEventListener('hashchange', () => {
+				willScroll = true;
+				activateTocLink(window.location.hash);
+			});
+
+			// watch for scroll
+			window.addEventListener(
+				'scroll',
+				() => {
+					// ignore hash-change-initiated scrolls
+					if (willScroll) {
+						willScroll = false;
+					} else {
+						activateTocLink(findActiveHeadingId());
+					}
+				},
+				{ passive: true }
+			);
+		}
+	})();
+</script>
diff --git a/components/MobileTableOfContents.astro b/components/MobileTableOfContents.astro
index 74342e3c5c3a2e52f6782846b8ff5547d7131d91..da2251222c7599ce2aec92b57233bd9d735d44c8 100644
--- a/components/MobileTableOfContents.astro
+++ b/components/MobileTableOfContents.astro
@@ -1,6 +1,6 @@
 ---
 import Icon from '../user-components/Icon.astro';
-import TableOfContentsList from './TableOfContents/TableOfContentsList.astro';
+import TableOfContentsList from './TableOfContentsList.astro';
 import type { Props } from '../props';
 
 const { labels, toc } = Astro.props;
@@ -108,15 +108,7 @@ const { labels, toc } = Astro.props;
 </style>
 
 <script>
-	import { StarlightTOC } from './TableOfContents/starlight-toc';
-
-	class MobileStarlightTOC extends StarlightTOC {
-		override set current(link: HTMLAnchorElement) {
-			super.current = link;
-			const display = this.querySelector('.display-current') as HTMLSpanElement;
-			if (display) display.textContent = link.textContent;
-		}
-
+	class MobileStarlightTOC extends HTMLElement {
 		constructor() {
 			super();
 			const details = this.querySelector('details');
diff --git a/components/Sidebar.astro b/components/Sidebar.astro
index 38dc898d3697b5b15f8e812711258757927d8088..4e42aec72f87c7139bae80c9f07a1f595dac1c78 100644
--- a/components/Sidebar.astro
+++ b/components/Sidebar.astro
@@ -11,3 +11,84 @@ const { sidebar } = Astro.props;
 <div class="md:sl-hidden">
 	<MobileMenuFooter {...Astro.props} />
 </div>
+
+{/* This is intentionally inlined to avoid FOUSC. */}
+<script is:inline>
+	(() => {
+		const sidebarContent = document.getElementById('starlight__sidebar')?.firstElementChild;
+
+		if (sidebarContent) {
+			// could have been as simple as
+			// sidebarContent.querySelectorAll('details:has(a[aria-current="page"])')
+			function findAllDetailsAncestors(activeLink, boundaryElement) {
+				const detailsElements = new Set();
+				let parent = activeLink.parentElement;
+				while (parent && parent !== boundaryElement) {
+					if (parent.tagName === 'DETAILS') {
+						detailsElements.add(parent);
+					}
+					parent = parent.parentElement;
+				}
+				return [...detailsElements];
+			}
+
+			// open ancestor details of the active item
+			const activeLink = sidebarContent.querySelector('a[aria-current="page"]');
+			const activeLinkDetailAncestors = findAllDetailsAncestors(activeLink, sidebarContent);
+
+			activeLinkDetailAncestors.forEach((detail) => {
+				detail.setAttribute('open', '');
+			});
+
+			const detailElements = [...sidebarContent.querySelectorAll('details')];
+
+			// Restore state from previous navigation
+			// (but keep the active item's ancestors open regardless)
+			const key = 'sl-sidebar-state';
+			const savedStateJson = sessionStorage.getItem(key);
+
+			if (savedStateJson) {
+				let savedState;
+
+				try {
+					savedState = JSON.parse(savedStateJson);
+				} catch (e) {
+					console.error('Error parsing saved position:', e);
+				}
+
+				if (savedState) {
+					savedState.details.forEach((isOpen, i) => {
+						if (isOpen) {
+							detailElements[i].setAttribute('open', '');
+						} else if (!activeLinkDetailAncestors.includes(detailElements[i])) {
+							detailElements[i].removeAttribute('open');
+						}
+					});
+
+					sidebarContent.scrollTop = savedState.scrollTop;
+				}
+			}
+
+			// Scroll the active list item into view if necessary
+			const activeLi = sidebarContent.querySelector('a[aria-current="page"]').parentElement;
+			if (activeLi) {
+				const sidebarRect = sidebarContent.getBoundingClientRect();
+				const liRect = activeLi.getBoundingClientRect();
+				if (liRect.top < sidebarRect.top || liRect.bottom > sidebarRect.bottom) {
+					activeLi.scrollIntoView({ behavior: 'instant', block: 'nearest' });
+				}
+			}
+
+			// Save user state before navigating
+			window.addEventListener('beforeunload', () => {
+				sessionStorage.setItem(
+					key,
+					JSON.stringify({
+						details: detailElements.map((detail) => detail.hasAttribute('open')),
+						scrollTop: sidebarContent.scrollTop
+					})
+				);
+			});
+		}
+	})();
+</script>
diff --git a/components/TableOfContents/starlight-toc.ts b/components/TableOfContents/starlight-toc.ts
deleted file mode 100644
index a5de15bd336c198fbc6b788b6351e2a3ae278dd4..0000000000000000000000000000000000000000
diff --git a/components/TableOfContents.astro b/components/TableOfContents.astro
index eafd3d868e9ae6965aecdbe117af1a218b174b93..1f774d2e76f7f6cf7b41765073df2d8d06d261ca 100644
--- a/components/TableOfContents.astro
+++ b/components/TableOfContents.astro
@@ -1,5 +1,5 @@
 ---
-import TableOfContentsList from './TableOfContents/TableOfContentsList.astro';
+import TableOfContentsList from './TableOfContentsList.astro';
 import type { Props } from '../props';
 
 const { labels, toc } = Astro.props;
@@ -7,7 +7,7 @@ const { labels, toc } = Astro.props;
 
 {
 	toc && (
-		<starlight-toc data-min-h={toc.minHeadingLevel} data-max-h={toc.maxHeadingLevel}>
+		<starlight-toc>
 			<nav aria-labelledby="starlight__on-this-page">
 				<h2 id="starlight__on-this-page">{labels['tableOfContents.onThisPage']}</h2>
 				<TableOfContentsList toc={toc.items} />
@@ -16,4 +16,15 @@ const { labels, toc } = Astro.props;
 	)
 }
 
-<script src="./TableOfContents/starlight-toc"></script>
+{/* See script in MarkdownContent.astro to set active item. */}
+
+{/* This must remain to avoid FOUC */}
+<script>
+	class StarlightTOC extends HTMLElement {
+		constructor() {
+			super();
+		}
+	}
+
+	customElements.define('starlight-toc', StarlightTOC);
+</script>
\ No newline at end of file
diff --git a/components/TableOfContents/TableOfContentsList.astro b/components/TableOfContentsList.astro
similarity index 100%
rename from components/TableOfContents/TableOfContentsList.astro
rename to components/TableOfContentsList.astro
diff --git a/components/ThemeSelect.astro b/components/ThemeSelect.astro
index de88627d2d6c556a1bc20e293463950537314ae8..989706dae9a0826c7c026b66d3382cd3619b9e72 100644
--- a/components/ThemeSelect.astro
+++ b/components/ThemeSelect.astro
@@ -40,6 +40,20 @@ const { labels } = Astro.props;
 				select.addEventListener('change', (e) => {
 					if (e.currentTarget instanceof HTMLSelectElement) {
 						this.#onThemeChange(this.#parseTheme(e.currentTarget.value));
+
+						// Fix for safari not redrawing the scroll bar until hover
+						// or a page refresh. Causes slight momentary layout shift,
+						// but that's less offensive than the wrong-color scrollbar.
+						const htmlElement = document.querySelector("html");
+						if (htmlElement) {
+							const initialOverflowY = window.getComputedStyle(htmlElement).overflowY;
+							requestAnimationFrame(() => {
+								document.documentElement.style.overflowY = "hidden";
+								requestAnimationFrame(() => {
+									document.documentElement.style.overflowY = initialOverflowY;
+								});
+							});
+						}
 					}
 				});
 			}